### PR TITLE
Fix MSAL cache cleaning method

### PR DESCRIPTION
**Pull request by pkanher617**
This PR is a fix for issue #527 where cache keys are not being cleared out. 

A previous PR (#544) was submitted for this, but was failing some use cases. This fix works for manually tested use cases and passes all unit tests.

### CODE REVIEW

File: lib/msal-core/src/Storage.ts [hunk 1]
```diff
@@ -90,18 +90,31 @@ export class Storage {// Singleton
...
+    removeAcquireTokenEntries(): void {
         const storage = window[this.cacheLocation];
         if (storage) {
             let key: string;
             for (key in storage) {
                 if (storage.hasOwnProperty(key)) {
-                    if ((authorityKey !== "" && key.indexOf(authorityKey) > -1) || (acquireTokenAccountKey !== "" && key.indexOf(acquireTokenAccountKey) > -1)) {
-                        this.removeItem(key);
+                    if (key.indexOf(CacheKeys.AUTHORITY) !== -1 || key.indexOf(CacheKeys.ACQUIRE_TOKEN_USER) !== 1) {
+                        const state = key.split(Constants.resourceDelimiter)[1];
```
Comment (DarylThayil) [inline lib/msal-core/src/Storage.ts]: any case where the array does not have element [1]

File: lib/msal-core/src/Storage.ts [hunk 2]
```diff
@@ -90,18 +90,31 @@ export class Storage {// Singleton
...
+                        if (state && !this.tokenRenewalInProgress(state)) {
```
Comment (DarylThayil) [inline lib/msal-core/src/Storage.ts]: what happens if a tokenRenewalInProgress?

**Comment by DarylThayil** [comment]
few questions / comments but looks good to me

File: lib/msal-core/src/UserAgentApplication.ts [hunk 3]
```diff
@@ -430,12 +430,7 @@ export class UserAgentApplication {
         request.state
       );
 
-      // Cache nonce
-      this.cacheStorage.setItem(Constants.nonceIdToken, serverAuthenticationRequest.nonce, this.inCookie);
-
-      // Cache account and authority
-      this.setAccountCache(account, serverAuthenticationRequest.state);
-      this.setAuthorityCache(serverAuthenticationRequest.state, acquireTokenAuthority.CanonicalAuthority);
+      this.updateAcquireTokenCacheEntries(serverAuthenticationRequest, account);
```
Comment (sameerag) [inline lib/msal-core/src/UserAgentApplication.ts]: @pkanher617 As discussed you may want to combine the login calls caching code for "authority" and "nonce" also into this API. Please factor that login calls do not have "acquireTokenAccountKey".

File: lib/msal-core/src/Storage.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-core/src/Storage.ts]: We always store CacheKeys.Authority or CacheKeys.AcquireToken with the resource delim and state. I can add a check though.

File: lib/msal-core/src/Storage.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-core/src/Storage.ts]: Then we don't remove the entries.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: added to updateCacheEntries. Please check new code.

File: lib/msal-core/src/Storage.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-core/src/Storage.ts]: updated

File: lib/msal-core/src/UserAgentApplication.ts [hunk 4]
```diff
@@ -674,13 +658,7 @@ export class UserAgentApplication {
         // populate QueryParameters (sid/login_hint/domain_hint) and any other extraQueryParameters set by the developer
         serverAuthenticationRequest = this.populateQueryParams(account, request, serverAuthenticationRequest);
 
-        // Cache nonce
-        this.cacheStorage.setItem(Constants.nonceIdToken, serverAuthenticationRequest.nonce, this.inCookie);
-        serverAuthenticationRequest.state = serverAuthenticationRequest.state;
-
-        // Cache account and authority
-        this.setAccountCache(account, serverAuthenticationRequest.state);
-        this.setAuthorityCache(serverAuthenticationRequest.state, acquireTokenAuthority.CanonicalAuthority);
```
Comment (manoj-rath) [inline lib/msal-core/src/UserAgentApplication.ts]: it's canonicalAuthority here, however in the updateCacheEntries is using authority

File: lib/msal-core/src/UserAgentApplication.ts [hunk 5]
```diff
@@ -430,12 +419,7 @@ export class UserAgentApplication {
         request.state
       );
 
-      // Cache nonce
-      this.cacheStorage.setItem(Constants.nonceIdToken, serverAuthenticationRequest.nonce, this.inCookie);
-
-      // Cache account and authority
-      this.setAccountCache(account, serverAuthenticationRequest.state);
-      this.setAuthorityCache(serverAuthenticationRequest.state, acquireTokenAuthority.CanonicalAuthority);
```
Comment (manoj-rath) [inline lib/msal-core/src/UserAgentApplication.ts]: CanonicalAuthority - same as the other comment

**Comment by DarylThayil** [comment]
lgtm approved

**Pull request merged**